01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+This is the case of logs/events in the EVM, which may be stored in a central EVM logs actor.
+
+## IPLD everything
+
+<!-- ipld-wasm library, codecs, selectors, IPLD code representation -->
+
+## State access
+
+<!-- avoiding excessive boundary costs, garbage collection -->
+
+## Chain access
+
+## Actor deployment
+
+The current InitActor (`f00`) will be extended with a (`LoadActor`) method that takes the actor's WASM bytecode as an input parameter and, potentially, the sandbox version (see above).
+
+It will:
+
+1. Validate the bytecode.
+    - Syntactical validation
+    - Potentially, structural validation.
+    - Potentially, lightweight static code analysis.
+2. Multihash the bytecode and calculate the CID.
+3. Check if the actor already exists in the `ActorRegistry`.
+4. If it exists, charge a small amount of gas and return the existing CID. If not, continue.
+5. Insert an entry in the `ActorRegistry`, binding the CID with the WASM bytecode.
+6. Return the CID and charge the full amount of gas for the length of the bytecode.
+
+At this point, the actor is ready to be instantiated through the standard `Exec` call of the `InitActor`. Any params provided will be passed through to the actor's constructor (method number = 0).
+
+## Cross-actor calls
+
+## Call patterns
+
+<!-- single entrypoint; sync only; future async; parallel execution; IPLD schemas -->
+
+## Syscalls
+
+## Gas accounting
+
+## WASM interpreters
+
+## JSON-RPC API
+
+## Interoperability with other networks
+
+## Formal verifiability
+
+## Upgradability
+
+<!-- at different layers, incl. evm shim -->
+
+## Annex: Current VM
+
+![Current VM architecture](img/current-vm-arch.png)
+
+The term _Actor_ is a reference to the [actor model](https://en.wikipedia.org/wiki/Actor_model), a concurrent computation paradigm that inspires Filecoin's runtime and scalability primitives.
+
+Actors operate on the state tree. Nothing else can modify the state tree in normal circumstances, other than actor logic. The single exception is state migration logic during a network upgrades. It can conduct bulk modifications, both to the content and the structure of the state tree.
+
+The state tree is an IPLD object, containing the root of an [HAMT](https://ipld.io/specs/advanced-data-layouts/hamt/spec/#appendix-filecoin-hamt-variant) which in turn contains all actors, keyed by ID address.
+
+Each actor has a type (represented by a CID), and a state root. The VM enforces actor state isolation, thus actors are prevented from accessing each other states.
+
+Message passing is used to communicate between actors, even when simple state accesses are required. Mutations to state may only be applied within transactions.
```

Comment:
I'd be honest here. We effectively use "RPC" calls, not message passing. We talk about it in terms of "send", but it synchronous and has a return value.